Process incoming transactions

torkelrogstad opened the pull request:
In this PR we

1) Add new tables for outgoing and incoming transaction outputs, and refactor what was previously called the UTXO table to be a spending info table:
    - incoming outputs have foreign keys to our addresses, on the SPK
    - incoming outputs have foreign keys to spending info rows, which contain information such as the output value, etc
    - outgoing outputs have foreign keys to incoming outputs 
2) Add SQL joins to find incoming outputs with their corresponding addresses
3) Make table creation on test startup be one SQL transaction, instead of many roundtrips
4) Add a method `HDChain.toAddress` in `core`
5) Remove all existing DAO fixtures for the wallet, and instead create one big fixture that creates all tables, drops all tables and provides all DAOs. The different DAO fixtures all did the same thing anyway, so there was no loss in removing them, really.
6) Replace `addUtxo` with `processTransaction` in the wallet API. `addUtxo` is still kept around as an internal utility, but not exposed to the user. `processTransaction` takes in a TX and a number of confirmations, and updates internal state accordingly. 
7) Add a method `getUnconfirmedBalance` to the wallet API. `getBalance` is also updated, taking advantage of our new DB tables. 


TODO: 
  - [x] Add foreign key to UTXOs in incoming TXs
  - [x] Make all DAO tests pass
 -  [x] Add `processTransaction` to wallet
 -  [x] Refactor incoming/outgong transaction model to work with incoming/outgoing TXOs

torkelrogstad committed:
Add SPK field to AddressTable

torkelrogstad committed:
Add tables for incoming and outgoing transactions

torkelrogstad committed:
Move utxos into wallet test util

torkelrogstad committed:
Create DBs transactionally

torkelrogstad committed:
Add toAddress on HDChain

torkelrogstad committed:
DB loggers exposed in logging configuration

torkelrogstad committed:
Add tx tables to WalletDbManagement

torkelrogstad committed:
Add foreign key on UTXOs to incoming transactions

torkelrogstad committed:
Generalize DAO fixture for wallet

torkelrogstad committed:
Move WalletDAOs to testkit

torkelrogstad committed:
Get UTXO DB unit tests passing

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/models/AddressTable.scala:
```diff
@@ -118,29 +136,27 @@ object AddressDbHelper {
+  def ecPublicKey: Rep[ECPublicKey] = column("pubkey")
 
-  def ecPublicKey: Rep[ECPublicKey] = column[ECPublicKey]("pubkey")
+  def hashedPubKey: Rep[Sha256Hash160Digest] = column("hashed_pubkey")
```
nit: but this probably should be changed to `hash` and be of type `Rep[HashDigest]`

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/models/TransactionTable.scala:
```diff
@@ -0,0 +1,92 @@
+/** Transactions our wallet has received */
+final case class IncomingTransaction(
+    transaction: Transaction,
+    scriptPubKey: ScriptPubKey,
```
i think output index is better than spk here, and we can add a method that gives amount, scriptPubKey, etc in the body of the case class

torkelrogstad replied:
The problem there is that we need to create a foreign key on `scriptPubKey` into address table. I don't know how to do that without having the actual SPK as a field in the case class...

Christewart replied:
```scala
require(transaction.outputs.exists(_.scriptPubKey == scriptPubKey))
```

torkelrogstad committed:
Add findTx to IncomingTxDAO

torkelrogstad committed:
Move ID in UTXO DB models for more ergonomic functions

torkelrogstad committed:
Replace addUtxo with processTransaction

In this commit we replace addUtxo with
processTransaction in the wallet API. addUtxo
is kept around, but only as an internal
implementation detail.

The new flow looks like this:
1) The wallet gets notified about a new TX
    through a processTransaction call
2) It then searches through our already
    processed incoming TXs:
        if found, updates the confirmation
            status
        else, inserts it into the DB
3) We then call addUtxo, if the previous
    step resulted in a new TX in our
    DB

torkelrogstad committed:
Add getUnconfirmedBalance to wallet

torkelrogstad committed:
Get rid of compiler warnings

torkelrogstad committed:
Update wallet integration tests

torkelrogstad committed:
Add ProcessTransactionTest

torkelrogstad committed:
Clean up Scaladoc and turn down logging

Christewart commented on core/src/main/scala/org/bitcoins/core/hd/HDChain.scala:
```diff
@@ -20,6 +20,9 @@ sealed abstract class HDChain extends BIP32Path {
   def toInt: Int = chainType.index
 
+  /** Given a index, creates a HD address */
+  def toAddress(index: Int): HDAddress = HDAddress(this, index = index)
```
suggestion: `toHDAddress`

Christewart commented on node-test/src/test/scala/org/bitcoins/node/NodeWithWalletTest.scala:
```diff
@@ -58,8 +60,16 @@ class NodeWithWalletTest extends BitcoinSWalletTest {
+              prevBalance <- wallet.getUnconfirmedBalance()
+              _ <- wallet.processTransaction(tx, confirmations = 0)
+              balance <- wallet.getUnconfirmedBalance()
+            } completionP.success(
```
This seems like weird syntax, why no yield here?

Christewart replied:
Also I can't get this test case to fail if I switch the predicate to be

```scala
assert(balance != prevBalance + amountFromBitcoind))
```

torkelrogstad replied:
Nice catch! You're right that this test was not falsifiable. This was because `assert` doesn't return anything if passed `false`, but throws a `TestFailedException`. The `.success(...)` method was therefore never evaluated. 

`completetionP.complete { Try { ... } } ` solved it!

Re. the syntax question: A `for` expression without a `yield` does a `foreach` instead on the final value instead of `map`. We don't use the end result for anything except for a side effect, so IMO not having a `yield` communicates intent better.

Christewart commented on testkit/src/main/scala/org/bitcoins/testkit/wallet/WalletTestUtil.scala:
```diff
@@ -64,5 +72,64 @@ object WalletTestUtil {
+
+    /** Get a TX with outputs */
+    def getTx: Transaction =
+      TransactionGenerators.transaction
```
It would probably be worth making a `TransactionGenerators.nonEmptyOutputs` generator

Christewart commented on wallet-test/src/test/scala/org/bitcoins/wallet/ProcessTransactionTest.scala:
```diff
@@ -0,0 +1,94 @@
+      oldConfirmed <- wallet.getBalance()
+      _ <- action // by name
+      newUtxos <- wallet.listUtxos()
+      newUnconfirmed <- wallet.getUnconfirmedBalance()
```
not used?

Christewart commented on wallet-test/src/test/scala/org/bitcoins/wallet/ProcessTransactionTest.scala:
```diff
@@ -0,0 +1,94 @@
+      val WalletWithBitcoind(wallet, bitcoind) = walletAndBitcoind
+
+      for {
+        unrelated <- bitcoind.getNewAddress
```
suggestion: Maybe used generators here so we don't need a full fledge bitcoind instance? It can be any arbitrary tx right?

Christewart commented on wallet-test/src/test/scala/org/bitcoins/wallet/WalletIntegrationTest.scala:
```diff
@@ -27,40 +27,45 @@ class WalletIntegrationTest extends BitcoinSWalletTest {
+      _ <- wallet.processTransaction(tx, confirmations = 0)
+
+      utxosPostAdd <- wallet.listUtxos()
+      _ = assert(utxosPostAdd.nonEmpty)
```
```scala
assert(utxosPostAdd.length == 1)
```

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -26,16 +26,34 @@ import org.bitcoins.wallet.ReadMnemonicError.JsonParsingError
+  /** Sums up the value of all incoming
+    * TXs in the wallet, filtered by the given predicate */
+  // TODO account for outgoing TXs
+  private def sumOfIncomingGiven(
```
`filterThenSum` as a name?

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,11 +182,160 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+    */
+  private def processNewIncomingTx(
+      transaction: Transaction,
+      confirmations: Int): Future[Option[IncomingTransaction]] = {
```
This return type should probably be `Future[Vector[IncomingTransaction]]`

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/models/TransactionTable.scala:
```diff
@@ -0,0 +1,106 @@
+}
+
+/** Transactions our wallet has received */
+final case class IncomingTransaction(
```
So I think this should be renamed to `IncomingUtxo`, as this has information relevant to the utxo not necessarily the tx itself. A `IncomingTransaction` shouldn't really be concerned about a `sriptPubKey`, `outputIdx` etc

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,11 +182,160 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+            s"Found $count relevant output(s) in transaction=${transaction.txIdBE}: $outputStr")
+          if (xs.length > 1) {
+            logger.warn(
+              s"${xs.length} SPKs were relevant to transaction=${transaction.txIdBE}, but we aren't able to handle more than 1 for the time being")
```
This needs a issue at minimum, it is not uncommon for people to re-use addresses (unfortunately) and this seems to break this.

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,11 +182,160 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+        }
+      }
+
+      relevantOutsWithIdx match {
```
Overally I think this should be moved into a function that processes each relevantOuts and returns a `Vector[Future[IncomingUtxo]]` and then sequences those Futures to return a single `Future[Vector[IncomingUtxo]]`

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,11 +182,160 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+          err))
+
+      updateF
+    } else if (foundTx.confirmations > foundTx.confirmations) {
```
This isn't satisfiable is it?

nkohen commented on testkit/src/main/scala/org/bitcoins/testkit/fixtures/WalletDAOFixture.scala:
```diff
@@ -0,0 +1,35 @@
+
+  def withFixture(test: OneArgAsyncTest): FutureOutcome =
+    makeFixture(build = () => WalletDbManagement.createAll().map(_ => daos),
+                destroy = () => WalletDbManagement.dropAll())(test)
```
I think it might be cleaner to decouple this from depending on `WalletDbManagement.allTables`, can we not pass in something like a `WalletDAOs.allTables` to be created so that we never have to worry about inconsistent states between the fixture and what is done by `WalletDbManagement`?

torkelrogstad replied:
Sorry, I don't get what you mean. Won't we have to have some sort of list of all tables regardless of what we end up doing? That's what `createAll` does, iterating over `allTables` and creating them

nkohen commented on testkit/src/main/scala/org/bitcoins/testkit/wallet/BitcoinSWalletTest.scala:
```diff
@@ -52,7 +52,8 @@ trait BitcoinSWalletTest
     WalletDbManagement
-      .dropAll()(config = config.walletConf, ec = implicitly[ExecutionContext])
+      .dropAll()(config = wallet.walletConfig,
+                 ec = implicitly[ExecutionContext])
```
If this gets resolved by `ExecutionContext.global` I think I'd prefer being explicit

torkelrogstad replied:
This doesn't resolve to `ExecutionContext.global` unless that is explicitly imported. 

For all def with implicit params:

```scala
def foo(bar: Int)(implicit s: Something): Unit = ???
```

All of the below snippets are equivalent: 

```scala
implicit val s: Something = ???
foo(2) // implicitly passed
```

```scala
val s: Something = ???
foo(2)(s) // explicitly passed
```

```scala
implicit val s: Something = ???
val otherS: Something = implicitly[Something] // resolves to s above
foo(2)(otherS)
```

```scala
implicit  val s: Something = ???
foo(2)(implicitly[Something] // this resolves to s above
```

There's no point in the last snippet (because `foo` only takes one implicit param). However on the line you commented on, there's two implicit params. So we're basically saying: "use this specific wallet config, but conjure up the `ExecutionContext` following normal implicit resolution rules"

Christewart replied:
I don't really understand what the benefit is of the extra complexity rather than just adding a implicit param on the `destroyWallet()` method. KISS

torkelrogstad replied:
We want to control precisely which wallet is being destroyed (which is achieved here by making sure we're operating on the correct datadir). If we're adding an implicit parameter to the method we're not getting any guarantees about what wallet we're destroying because it's just grabbing whatever implicit configuration is closest in scope.

nkohen commented on wallet-test/src/test/scala/org/bitcoins/wallet/models/UTXOSpendingInfoDAOTest.scala:
```diff
@@ -5,45 +5,57 @@ import org.bitcoins.core.protocol.transaction.{
+          val output = TransactionOutput(1.bitcoin, WalletTestUtil.sampleSPK)
+          val scriptWitness = WalletTestUtil.sampleScriptWitness
+          val privkeyPath = WalletTestUtil.sampleSegwitPath
+          NativeV0UTXOSpendingInfoDb(outPoint = outpoint,
```
any reason not to just put these outside of the block they're in and have this be `val utxo`?

nkohen replied:
same in other places, obvs a nit

nkohen commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -26,16 +26,34 @@ import org.bitcoins.wallet.ReadMnemonicError.JsonParsingError
+
+  /** Sums up the value of all incoming
+    * TXs in the wallet, filtered by the given predicate */
+  // TODO account for outgoing TXs
```
issue for this?

torkelrogstad replied:
Added it to the umbrella issue: https://github.com/bitcoin-s/bitcoin-s/issues/530

nkohen commented on wallet/src/main/scala/org/bitcoins/wallet/models/TransactionTable.scala:
```diff
@@ -0,0 +1,106 @@
+
+  // TODO: What happens if we get paid to multiple SPKs in the same
+  // transaction? Need to make a table of SPKs, and map IDs in that
+  // table to TXs in this table...
```
issue

torkelrogstad committed:
Add TX generators that pay to a given SPK

torkelrogstad committed:
Replace incoming/outgoing TXs with TXOs

Rework the wallet DB model so that instead
of dealing with incoming and outgoing transactions
we deal with incoming and outgoing transaction
outputs.

torkelrogstad committed:
Fix bad comparison in ProcessTransactionTest

torkelrogstad committed:
Rename internal helper method in LockedWallet

torkelrogstad committed:
Rename HDCHain.toAddress

torkelrogstad committed:
Add sampleSome method to scalacheck.Gen

torkelrogstad committed:
Add TX generator for non-empty output TXs

torkelrogstad committed:
Make NodeWithWalletTest properly falsifiable

torkelrogstad committed:
Rebase cleanup

Christewart commented on testkit/src/main/scala/org/bitcoins/testkit/Implicits.scala:
```diff
@@ -0,0 +1,23 @@
+
+    @tailrec
+    /** Gets a sample from this generator that's not `None` */
+    def sampleSome: T = gen.sample match {
```
Just a note, this can really slow down tests if there is a complex generator being used

Christewart replied:
This seems like the worse of both worlds, the type `Gen` is around for a reason, is there a reason why we aren't doing things properly and just writing a property based test instead of unit tests with one random generated value?

torkelrogstad replied:
This is meant as a better alternative than `FooGenerator.sample.get`. In many tests in the code base we don't want to have a full blown property-based test (for example while doing wallet tests, running the same test 10 times with marginally different inputs would slow things down _a lot_), but we need testing data that is actually meaningful (like for example a transaction paying to a specific SPK)

Christewart replied:
This can also slow down the code significantly because it is guaranteed that we generate a value. This can essentially end up in an infinite loop trying to satisfy the generator. That is why the `Option` type is returned.

Christewart replied:
I think we should just cache these values in a `*TestUtil.scala` file if you don't want to do the full blow property based testing.

torkelrogstad replied:
If we have a generator that leads to a significant slowdown/infinite loop I'd say that's a problem on its own, as that's going to make any tests using that generator equally slow. The solution in that case is fixing the generator. 

For a bunch of things, having a `"TestUtil` object works fine, but what about for example an address that pays to a specific SPK, and that SPK is generated by the wallet in a test? I'm thinking `sampleSome` works fine in that case. 

If you'd like, I could add a constraint to the method that it's recursing maximum 10 levels deep, is that something you could live with?

Christewart replied:
If you feel strongly about it I will let it go, but I think this is a classic example of circumventing the type system safety to be more convenient. Circumventing the type system is more acceptable in test code, but I do think it will bite us in the long term since the library designer designed the API like that for a reason.

Christewart commented on testkit/src/main/scala/org/bitcoins/testkit/core/gen/TransactionGenerators.scala:
```diff
@@ -132,14 +147,23 @@ trait TransactionGenerators extends BitcoinSLogger {
-  /** Generates a random [[org.bitcoins.core.protocol.transaction.WitnessTransaction WitnessTransaction]] */
-  def witnessTransaction: Gen[WitnessTransaction] =
+  /** Generates a legacy transaction with at least one output paying to the given SPK */
+  def baseTransactionTo(spk: ScriptPubKey) =
```
nit: Add return type parameter

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,10 +182,186 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+      logger.warn(msg)
+      Future.failed(new RuntimeException(msg))
+    } else {
+      logger.debug(
```
This could be bad in the case of competing re-orgs. Do we really want this to be our behavior?

torkelrogstad replied:
Reorg isn't handled any where in the wallet currently. I can add a comment here, saying that this will need to change once reorg handling is implemented.

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,10 +182,186 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+        .findTx(transaction)
+        .flatMap {
+          // no existing elements found
+          case Vector() =>
```
`Vector.empty`

torkelrogstad replied:
`Vector.empty` is a `def` so it cannot be used in pattern-matching :-(

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,10 +182,186 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+          case txos: Vector[IncomingWalletTXO] =>
+            Future
+              .sequence {
+                txos
```
This is pretty hairy

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,10 +182,186 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+              s"${xs.length} SPKs were relevant to transaction=${transaction.txIdBE}, but we aren't able to handle more than 1 for the time being")
+          }
+
+          def processUtxo(index: Int): Future[SpendingInfoDb] =
```
I find methods like these very hard to read when they are nested inside of a very large method _and_ they don't use `{}` braces explicitly. The same thing applies with `createIncomingTxo`

torkelrogstad committed:
Refactor hairy expression

torkelrogstad committed:
Add annotated type to public method

Christewart commented on wallet/src/main/scala/org/bitcoins/wallet/LockedWallet.scala:
```diff
@@ -161,10 +182,186 @@ abstract class LockedWallet extends LockedWalletApi with BitcoinSLogger {
+      logger.debug(
+        s"Increasing confirmation count of txo=${transaction.txIdBE}, old=${foundTxo.confirmations} new=${confirmations}")
+      val updateF =
+        incomingTxoDAO.update(foundTxo.copy(confirmations = confirmations))
```
I don't think we should be making the assumption that this is still in the same transaction. Someone could be trying to double spend. If I understand this code correctly, we are assuming everything else is still the same about the tx and it is a simple reorg?

torkelrogstad committed:
Refactor TX processing logic and add comments re. TODOs

torkelrogstad committed:
Have sampleSome do a maximum of 10 attempts

torkelrogstad committed:
Remove bad import
